corehq/apps/locations/models.py

```diff
@@ -9,20 +9,85 @@
 import json_field
 from casexml.apps.case.cleanup import close_case
 from corehq.apps.commtrack.const import COMMTRACK_USERNAME
-from corehq.apps.domain.models import Domain
 from corehq.apps.products.models import SQLProduct
 from mptt.models import MPTTModel, TreeForeignKey
-from corehq.apps.domain.models import Domain
+
 
 LOCATION_SHARING_PREFIX = 'locationgroup-'
 LOCATION_REPORTING_PREFIX = 'locationreportinggroup-'
 
 
+class LocationTypeManager(models.Manager):
+    def full_heirarchy(self, domain):
+        """
+        Returns a graph of the form
+        {
+           '<loc_type_id>: (
+               loc_type,
+               {'<child_loc_type_id>': (child_loc_type, [...])}
+           )
+        }
+        """
+        heirarchy = {}
+
+        def insert_loc_type(loc_type):
+            """
+            Get parent location's heirarchy, insert loc_type into it, and return
+            heirarchy below loc_type
+            """
+            if not loc_type.parent_type:
+                lt_heirarchy = heirarchy
+            else:
+                lt_heirarchy = insert_loc_type(loc_type.parent_type)
+            if loc_type.id not in lt_heirarchy:
+                lt_heirarchy[loc_type.id] = (loc_type, {})
+            return lt_heirarchy[loc_type.id][1]
+
+        for loc_type in self.filter(domain=domain).all():
+            insert_loc_type(loc_type)
+
+        return heirarchy
+
+    def by_domain(self, domain):
+        """
+        Sorts location types by heirarchy
```

Comment:
yeah, it's like depth-first search.  I'm using this ordering on the edit location types page, before I added it, they just appeared in an arbitrary order, which looks confusing when there's a strict ~~heirarchy~~ hierarchy implied.  This is arguably unneeded extra computation if you needed to get all LocationTypes and don't care about order, but I don't think that's something that really happens except when you're going to display them to the user.